### PR TITLE
294 fallbackroute should return opt

### PR DESCRIPTION
**Pull request by ekondrashev**
Please review this PR for #294 

It switches `Fallback.route` to use `Opt` instead of `Iterator`

### CODE REVIEW

**Comment by davvd** [comment]
@ekondrashev Thanks for your pull request, let me find someone who can review it

**Comment by davvd** [comment]
@pinaf this pull request is for you, please review

**Comment by ekondrashev** [comment]
@pinaf seems like the PR was lost, could you please provide your comments?

File: src/main/java/org/takes/facets/fallback/FbStatus.java [hunk 1]
```diff
@@ -96,20 +93,20 @@ public boolean fits(final Integer status) {
...
-                    final Collection<Response> rsp = new ArrayList<Response>(1);
+                    final Opt<Response> rsp;
                     if (check.fits(req.code())) {
-                        final Iterator<Response> iter = fallback.route(req);
-                        if (iter.hasNext()) {
-                            rsp.add(iter.next());
-                        }
+                        rsp = fallback.route(req);
+                    } else {
+                        rsp = new Opt.Empty<Response>();
```
Comment (pinaf) [inline src/main/java/org/takes/facets/fallback/FbStatus.java]: @ekondrashev move this assignment to line 103 above.

File: src/test/java/org/takes/facets/fallback/TkFallbackTest.java [hunk 2]
```diff
@@ -60,10 +59,10 @@ public void fallsBack() throws IOException {
                     new TkFailure(err),
                     new Fallback() {
                         @Override
-                        public Iterator<Response> route(final RqFallback req) {
-                            return Collections.<Response>singleton(
-                                new RsText(req.throwable().getMessage())
-                            ).iterator();
+                        public Opt<Response> route(final RqFallback req) {
+                            return new Opt.Single<Response>(
+                                    new RsText(req.throwable().getMessage())
```
Comment (pinaf) [inline src/test/java/org/takes/facets/fallback/TkFallbackTest.java]: @ekondrashev too much indentation here

**Comment by pinaf** [comment]
@ekondrashev 2 minor comments above

**Comment by pinaf** [comment]
@ekondrashev we need a merge with master

**Comment by ekondrashev** [comment]
@pinaf I've merged with master and fixed all the comments.
I've also added ClassDataAbstractionCouplingCheck check style warning suppression for FbStatus.java not sure if I need to create a ticket to resolve it as puzzle.

Please let me know if PR needs another changes.

File: .rultor.yml [hunk 3]
```diff
@@ -34,7 +34,9 @@ merge:
     mvn clean
     pdd --source=$(pwd) --verbose --file=/dev/null
   commanders:
+  - caarlos0
```
Comment (pinaf) [inline .rultor.yml]: @ekondrashev this doesn't look good.. looks like your merge was not done right

File: README.md [hunk 4]
```diff
@@ -419,7 +420,7 @@ new TkFork(
 Here is how you can parse an instance of `Request`:
 
 ```java
-Href href = new RqHref(request).href();
+Href href = new RqHref.Base(request).href();
```
Comment (pinaf) [inline README.md]: @ekondrashev this doesn't look good.. looks like your merge was not done right

**Comment by pinaf** [comment]
@ekondrashev looks like your merge was not done correctly. there are multiple files that should not have any changes but do..

**Comment by ekondrashev** [comment]
@pinaf sorry for that, fixed.

File: src/main/java/org/takes/facets/fallback/FbStatus.java [hunk 5]
```diff
@@ -43,8 +40,10 @@
  * <p>The class is immutable and thread-safe.
  *
  * @author Yegor Bugayenko (yegor@teamed.io)
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
```
Comment (pinaf) [inline src/main/java/org/takes/facets/fallback/FbStatus.java]: @ekondrashev this is duplicated

File: src/main/java/org/takes/facets/fallback/FbStatus.java [hunk 6]
```diff
@@ -43,8 +40,10 @@
...
  * @version $Id$
  * @since 0.13
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
```
Comment (pinaf) [inline src/main/java/org/takes/facets/fallback/FbStatus.java]: @ekondrashev are you sure we even need this suppression? it looks to me that you actually reduced the coupling (the imports above were reduced)

**Comment by pinaf** [comment]
@ekondrashev thanks. 2 comments

**Comment by ekondrashev** [comment]
@pinaf removed the suppression. As for the second comment, yes it is needed, qulice gives 
```
[ERROR] /src/main/java/org/takes/facets/fallback/FbStatus.java[46]: Class Data Abstraction Coupling is 8 (max allowed is 7) classes [Condition, Fallback, Opt.Empty, Opt.Single, RsWithBody, RsWithStatus, RsWithType, TkFixed]. (ClassDataAbstractionCouplingCheck)
``` 
without it.

**Comment by pinaf** [comment]
@ekondrashev thanks

**Comment by pinaf** [comment]
@rultor merge

File: src/main/java/org/takes/facets/fallback/FbChain.java [hunk 7]
```diff
@@ -55,21 +53,22 @@ public FbChain(final Fallback... fallbacks) {
...
                     throws IOException {
-                    final Collection<Response> rsp = new ArrayList<Response>(1);
+                    Opt<Response> rsp = new Opt.Empty<Response>();
                     for (final Fallback fbk : fallbacks) {
-                        final Iterator<Response> iter = fbk.route(req);
-                        if (iter.hasNext()) {
-                            rsp.add(iter.next());
+                        final Opt<Response> opt = fbk.route(req);
+                        if (opt.has()) {
+                            rsp = new Opt.Single<Response>(opt.get());
```
Comment (yegor256) [inline src/main/java/org/takes/facets/fallback/FbChain.java]: @ekondrashev why not just `rsp = opt`?

**Comment by yegor256** [comment]
@ekondrashev one comment above

**Comment by ekondrashev** [comment]
@yegor256 right, fixed

**Comment by ekondrashev** [comment]
@yegor256 could you please take a look at this PR?

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by ekondrashev** [comment]
@yegor256 could you please give an advice on the error above?

**Comment by yegor256** [comment]
@ekondrashev we tried to fix that in #309, but failed, as I see...

**Comment by yegor256** [comment]
@rultor try to merge again

**Comment by ekondrashev** [comment]
@yegor256 can we perhaps disable the test and reopen the #309?

**Comment by yegor256** [comment]
@ekondrashev #309 is already closed and paid, no way to reopen it. But you can report a new ticket, explain the problem there and wait here until it's fixed

**Comment by ekondrashev** [comment]
@yegor256 done, plz see #333 
@davvd this PR merge is blocked by #333 hence original defect is also blocked, will comment on it also

**Comment by pinaf** [comment]
@davvd we're waiting for #333 here

**Comment by pinaf** [comment]
@yegor256 since the test is unstable, we could try to merge again I supposed. Maybe we get lucky this time :)

**Comment by yegor256** [comment]
@rultor try to merge again

**Comment by davvd** [comment]
> @davvd we're waiting for #333 here

@pinaf right, let's wait for #333

**Comment by ekondrashev** [comment]
@davvd I believe there is no reason to wait for #333 now, since the PR was merged and initial ticket is closed already.

**Comment by davvd** [comment]
@elenavolokhova please, review this ticket for compliance with our [QA rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor please deploy

**Comment by elenavolokhova** [comment]
@davvd Quality is good here.

**Comment by davvd** [comment]
> @davvd Quality is good here.

@elenavolokhova thanks for the review and a good remark

**Comment by davvd** [comment]
@pinaf since quality is good, 10 mins paid to @yegor256 (our architect), transaction `000-6a463b2b`

**Comment by davvd** [comment]
@pinaf 10 mins was added to the account of @elenavolokhova (for QA review), in transaction 58704981; Thanks a lot, I just topped your account for **23 mins**, transaction ID `AP-29N57172GJ6916946`, total time was 426 hours and 41 mins; extra minutes for review comments (c=8); +23 added to your rating, current score is: [+7665](http://www.netbout.com/b/35128?open=rating)

**Pull request merged**